sources/packages/backend/libs/services/src/msfaa-number/msfaa-number-shared.service.ts

```diff
@@ -136,19 +181,51 @@ export class MSFAANumberSharedService {
           updatedAt: now,
         },
       );
-      // Create the new MSFAA record for the student.
-      const newMSFAANumber = new MSFAANumber();
-      newMSFAANumber.msfaaNumber = await this.consumeNextSequence(
-        offeringIntensity,
-      );
-      newMSFAANumber.student = { id: studentId } as Student;
-      newMSFAANumber.referenceApplication = {
-        id: referenceApplicationId,
-      } as Application;
-      newMSFAANumber.offeringIntensity = offeringIntensity;
-      newMSFAANumber.creator = { id: auditUserId } as User;
-      newMSFAANumber.createdAt = now;
-      await entityManager.getRepository(MSFAANumber).save(newMSFAANumber);
+      // The reactivated msfaaNumber.
+      let newMSFAANumber: MSFAANumber = {
+        id: options?.msfaaNumberId,
+      } as MSFAANumber;
+      // If there is not a MSFAA number to be re-activated.
+      if (!options?.msfaaNumberId) {
+        // Create the new MSFAA record for the student.
+        newMSFAANumber = new MSFAANumber();
+        newMSFAANumber.msfaaNumber = await this.consumeNextSequence(
+          offeringIntensity,
+        );
+        newMSFAANumber.student = { id: studentId } as Student;
+        newMSFAANumber.referenceApplication = {
+          id: referenceApplicationId,
+        } as Application;
+        newMSFAANumber.offeringIntensity = offeringIntensity;
+        newMSFAANumber.creator = { id: auditUserId } as User;
+        newMSFAANumber.createdAt = now;
+        await entityManager.getRepository(MSFAANumber).save(newMSFAANumber);
+      } else {
+        // Reactivate this msfaa record.
+        const systemUser = await this.systemUsersService.systemUser();
+        const updateResult = await this.dataSource
+          .getRepository(MSFAANumber)
+          .update(
+            {
+              id: options?.msfaaNumberId,
+            },
+            {
+              dateSigned: getISODateOnlyString(options?.dateSigned),
+              serviceProviderReceivedDate: getISODateOnlyString(
+                options?.serviceProviderReceivedDate,
+              ),
+              cancelledDate: null,
+              newIssuingProvince: null,
+              modifier: systemUser,
+            },
+          );
+        // Incase no record updated.
+        if (!updateResult.affected) {
+          throw new Error(
+            `Error while updating MSFAA number: ${options?.msfaaNumber}. Number of affected rows was ${updateResult.affected}, expected 1.`,
```

Comment:
The MSFAA number is passed through the methods till here and the only intention would be to generate this log. I am not sure if it is worth it. This method can throw the error message with the information that it has.

Reply:
IMO throwing error line 223-227 is not required here as MSFAA is validated preciously when we check for cancelled date is not null.

Reply:
IMO the method must that it was executed as expected independently of what the consumer is validating.
The update here is critical to ensure that the next update to have the MSFAA association will happen as expected.
I do not see a problem to have this check in place to ensure the method's consistency.
We do not validate for every single service method every single condition/parameter, but I see value in having it here also. Pease let me know if we need further discussion on this 😉

Reply:
I am good.